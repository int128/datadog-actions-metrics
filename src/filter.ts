import { minimatch } from 'minimatch'
import { v1 } from '@datadog/datadog-api-client'

type Inputs = {
  metricsPatterns: string[]
  datadogTags: string[]
}

export type MetricsFilter = <S extends v1.Series | v1.DistributionPointsSeries>(series: S[]) => S[]

export const createMetricsFilter = (inputs: Inputs): MetricsFilter => {
  const matcher = createMatcher(inputs.metricsPatterns)
  return (series) => {
    series = series.filter((s) => matcher(s.metric))
    return injectTags(series, inputs.datadogTags)
  }
}

export const createMatcher =
  (patterns: string[]) =>
  (metric: string): boolean => {
    if (patterns.length === 0) {
      return true
    }
    let matched = false
    for (const pattern of patterns) {
      if (pattern.startsWith('!')) {
        matched = matched && minimatch(metric, pattern)
      } else {
        matched = matched || minimatch(metric, pattern)
      }
    }
    return matched
  }

export const injectTags = <S extends { tags?: string[] }>(series: S[], tags: string[]): S[] => {
  if (tags.length === 0) {
    return series
  }
  return series.map((s) => ({ ...s, tags: [...(s.tags ?? []), ...tags] }))
}
