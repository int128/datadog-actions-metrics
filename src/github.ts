import assert from 'assert'
import * as fs from 'fs/promises'
import { Octokit } from '@octokit/action'
import { WebhookEvent } from '@octokit/webhooks-types'
import { retry } from '@octokit/plugin-retry'

export const getOctokit = (token: string) => new (Octokit.plugin(retry))({ auth: token, authStrategy: null })

export type Context = {
  repo: {
    owner: string
    repo: string
  }
  eventName: string
  payload: WebhookEvent
}

export const getContext = async (): Promise<Context> => {
  // https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables
  return {
    repo: getRepo(),
    eventName: getEnv('GITHUB_EVENT_NAME'),
    payload: JSON.parse(await fs.readFile(getEnv('GITHUB_EVENT_PATH'), 'utf-8')) as WebhookEvent,
  }
}

const getRepo = () => {
  const [owner, repo] = getEnv('GITHUB_REPOSITORY').split('/')
  return { owner, repo }
}

const getEnv = (name: string): string => {
  assert(process.env[name], `${name} is required`)
  return process.env[name]
}
