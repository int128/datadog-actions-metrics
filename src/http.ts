import { HttpClient, HttpClientResponse } from '@actions/http-client'
import { client } from '@datadog/datadog-api-client'
import { Readable } from 'stream'

export class HttpLibrary implements client.HttpLibrary {
  private client = new HttpClient()

  async send(req: client.RequestContext): Promise<client.ResponseContext> {
    const resp = await this.client.request(req.getHttpMethod(), req.getUrl(), toRequestData(req), req.getHeaders())
    return new client.ResponseContext(resp.message.statusCode ?? 0, toResponseHeaders(resp), {
      text: async () => await resp.readBody(),
      binary: async () => {
        const buffers = []
        for await (const data of resp.message) {
          buffers.push(data)
        }
        return Buffer.concat(buffers)
      },
    })
  }
}

const toRequestData = (req: client.RequestContext): string | Readable | null => {
  const body = req.getBody()
  if (body instanceof Buffer) {
    return Readable.from(body)
  }
  return body ?? null
}

const toResponseHeaders = (resp: HttpClientResponse): { [key: string]: string } => {
  const headers: { [key: string]: string } = {}
  for (const [k, v] of Object.entries(resp.message.headers)) {
    if (typeof v === 'string') {
      headers[k] = v
    } else if (v !== undefined) {
      headers[k] = v[0]
    }
  }
  return headers
}
