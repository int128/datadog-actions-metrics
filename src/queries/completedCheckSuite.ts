import { CompletedCheckSuiteQuery, CompletedCheckSuiteQueryVariables } from '../generated/graphql'
import { CheckAnnotation, CheckRun, CheckStep } from '../generated/graphql-types'
import { Octokit } from '../types'

const query = /* GraphQL */ `
  query completedCheckSuite($node_id: ID!, $workflow_path: String!) {
    node(id: $node_id) {
      __typename
      ... on CheckSuite {
        checkRuns(first: 100, filterBy: { checkType: LATEST }) {
          nodes {
            databaseId
            name
            startedAt
            completedAt
            conclusion
            status
            annotations(first: 10) {
              nodes {
                message
              }
            }
            steps(first: 50) {
              nodes {
                number
                name
                conclusion
                status
                startedAt
                completedAt
              }
            }
          }
        }
        commit {
          file(path: $workflow_path) {
            object {
              __typename
              ... on Blob {
                text
              }
            }
          }
        }
      }
    }
  }
`

type NonNullablePick<T, K extends keyof T> = {
  [P in K]-?: NonNullable<T[P]>
}

export type CompletedCheckSuite = {
  node: {
    __typename: 'CheckSuite'
    checkRuns: {
      nodes: CompletedCheckRun[]
    }
    commit: {
      file: {
        object: {
          text: string
        }
      }
    }
  }
}

type CompletedCheckRun = Pick<CheckRun, 'databaseId' | 'name' | 'status'> &
  NonNullablePick<CheckRun, 'startedAt' | 'completedAt' | 'conclusion'> & {
    annotations: {
      nodes: Pick<CheckAnnotation, 'message'>[]
    }
    steps: {
      nodes: CompletedStep[]
    }
  }

type CompletedStep = Pick<CheckStep, 'number' | 'name' | 'status'> &
  NonNullablePick<CheckStep, 'startedAt' | 'completedAt' | 'conclusion'>

export const queryCompletedCheckSuite = async (
  o: Octokit,
  v: CompletedCheckSuiteQueryVariables
): Promise<CompletedCheckSuite | undefined> => {
  const r = await o.graphql<CompletedCheckSuiteQuery>(query, v)
  return {
    node: {
      __typename: 'CheckSuite',
      checkRuns: extractCheckRuns(r),
      commit: extractCommit(r),
    },
  }
}

const extractCheckRuns = (r: CompletedCheckSuiteQuery): CompletedCheckSuite['node']['checkRuns'] => {
  if (r.node?.__typename !== 'CheckSuite') {
    throw new Error(`invalid __typename ${String(r.node?.__typename)} !== CheckSuite`)
  }

  const checkRuns: CompletedCheckRun[] = []
  for (const checkRun of r.node.checkRuns?.nodes ?? []) {
    if (checkRun == null) {
      continue
    }

    const annotations = []
    for (const annotation of checkRun.annotations?.nodes ?? []) {
      if (annotation == null) {
        continue
      }
      annotations.push(annotation)
    }

    const steps: CompletedStep[] = []
    for (const step of checkRun.steps?.nodes ?? []) {
      if (step == null) {
        continue
      }
      const { startedAt, completedAt, conclusion } = step
      if (startedAt == null || completedAt == null || conclusion == null) {
        continue
      }
      steps.push({
        ...step,
        startedAt,
        completedAt,
        conclusion,
      })
    }

    const { startedAt, completedAt, conclusion } = checkRun
    if (startedAt == null || completedAt == null || conclusion == null) {
      continue
    }
    checkRuns.push({
      ...checkRun,
      startedAt,
      completedAt,
      conclusion,
      annotations: { nodes: annotations },
      steps: { nodes: steps },
    })
  }
  return { nodes: checkRuns }
}

const extractCommit = (r: CompletedCheckSuiteQuery): CompletedCheckSuite['node']['commit'] => {
  if (r.node?.__typename !== 'CheckSuite') {
    throw new Error(`invalid __typename ${String(r.node?.__typename)} !== CheckSuite`)
  }
  if (r.node.commit.file?.object?.__typename !== 'Blob') {
    throw new Error(`invalid __typename ${String(r.node.commit.file?.object?.__typename)} !== Blob`)
  }
  const { text } = r.node.commit.file.object
  if (text == null) {
    throw new Error(`invalid text ${String(text)}`)
  }
  return {
    file: {
      object: {
        text,
      },
    },
  }
}
