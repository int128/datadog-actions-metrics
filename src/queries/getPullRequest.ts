import assert from 'assert'
import { GetPullRequestQuery, GetPullRequestQueryVariables } from '../generated/graphql.js'
import { Octokit } from '../types.js'

const query = /* GraphQL */ `
  query getPullRequest($owner: String!, $name: String!, $number: Int!) {
    repository(owner: $owner, name: $name) {
      pullRequest(number: $number) {
        commits(first: 1) {
          nodes {
            commit {
              authoredDate
              committedDate
            }
          }
        }
      }
    }
  }
`

export type PullRequestFirstCommit = {
  authoredDate: string
  committedDate: string
}

export const getPullRequestFirstCommit = async (
  o: Octokit,
  v: GetPullRequestQueryVariables,
): Promise<PullRequestFirstCommit | undefined> => {
  const r = await o.graphql<GetPullRequestQuery>(query, v)
  assert(r.repository != null)
  assert(r.repository.pullRequest != null)
  assert(r.repository.pullRequest.commits != null)
  assert(r.repository.pullRequest.commits.nodes != null)
  if (r.repository.pullRequest.commits.nodes.length === 0) {
    return // pull request has no commit
  }
  assert(r.repository.pullRequest.commits.nodes[0] != null)
  return r.repository.pullRequest.commits.nodes[0].commit
}
