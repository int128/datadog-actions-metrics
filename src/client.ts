import * as core from '@actions/core'
import { client, v1 } from '@datadog/datadog-api-client'
import { HttpLibrary } from './http.js'
import { createMetricsFilter, MetricsFilter } from './filter.js'

type Inputs = {
  datadogApiKey?: string
  datadogSite?: string
  datadogTags: string[]
  metricsPatterns: string[]
}

export type MetricsClient = {
  submitMetrics: (series: v1.Series[], description: string) => Promise<void>
  submitDistributionPoints(series: v1.DistributionPointsSeries[], description: string): Promise<void>
}

class DryRunMetricsClient implements MetricsClient {
  constructor(private readonly metricsFilter: MetricsFilter) {}

  // eslint-disable-next-line @typescript-eslint/require-await
  async submitMetrics(series: v1.Series[], description: string): Promise<void> {
    series = this.metricsFilter(series)
    core.startGroup(`Metrics payload (dry-run) (${description})`)
    core.info(JSON.stringify(series, undefined, 2))
    core.endGroup()
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  async submitDistributionPoints(series: v1.DistributionPointsSeries[], description: string): Promise<void> {
    series = this.metricsFilter(series)
    core.startGroup(`Distribution points payload (dry-run) (${description})`)
    core.info(JSON.stringify(series, undefined, 2))
    core.endGroup()
  }
}

class RealMetricsClient implements MetricsClient {
  constructor(
    private readonly metricsApi: v1.MetricsApi,
    private readonly metricsFilter: MetricsFilter,
  ) {}

  async submitMetrics(series: v1.Series[], description: string): Promise<void> {
    series = this.metricsFilter(series)
    core.startGroup(`Metrics payload (${description})`)
    core.info(JSON.stringify(series, undefined, 2))
    core.endGroup()
    core.info(`Sending ${series.length} metrics to Datadog`)
    const accepted = await this.metricsApi.submitMetrics({ body: { series } })
    core.info(`Sent ${JSON.stringify(accepted)}`)
  }

  async submitDistributionPoints(series: v1.DistributionPointsSeries[], description: string): Promise<void> {
    series = this.metricsFilter(series)
    core.startGroup(`Distribution points payload (${description})`)
    core.info(JSON.stringify(series, undefined, 2))
    core.endGroup()
    core.info(`Sending ${series.length} distribution points to Datadog`)
    const accepted = await this.metricsApi.submitDistributionPoints({ body: { series } })
    core.info(`Sent ${JSON.stringify(accepted)}`)
  }
}

export const createMetricsClient = (inputs: Inputs): MetricsClient => {
  if (inputs.datadogApiKey === undefined) {
    return new DryRunMetricsClient(createMetricsFilter(inputs))
  }

  const configuration = client.createConfiguration({
    authMethods: {
      apiKeyAuth: inputs.datadogApiKey,
    },
    httpApi: createHttpLibraryIfHttpsProxy(),
  })
  if (inputs.datadogSite) {
    client.setServerVariables(configuration, {
      site: inputs.datadogSite,
    })
  }
  return new RealMetricsClient(new v1.MetricsApi(configuration), createMetricsFilter(inputs))
}

const createHttpLibraryIfHttpsProxy = () => {
  const httpsProxy = process.env['https_proxy']
  if (httpsProxy) {
    core.info(`Using https_proxy: ${httpsProxy}`)
    return new HttpLibrary()
  }
}
