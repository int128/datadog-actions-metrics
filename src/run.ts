import * as core from '@actions/core'
import * as github from '@actions/github'
import { PullRequestEvent, PushEvent, WorkflowRunEvent } from '@octokit/webhooks-types'
import { computeRateLimitMetrics } from './rateLimit/metrics'
import { GitHubContext } from './types'
import { MetricsClient, createMetricsClient } from './client'
import { handleWorkflowRun } from './workflowRun/handler'
import { handlePullRequest } from './pullRequest/handler'
import { handlePush } from './push/handler'
import { handleSchedule } from './schedule/handler'

type Inputs = {
  githubToken: string
  githubTokenForRateLimitMetrics: string
  datadogApiKey?: string
  datadogSite?: string
  datadogTags: string[]
  collectJobMetrics: boolean
  collectStepMetrics: boolean
  preferDistributionWorkflowRunMetrics: boolean
  preferDistributionJobMetrics: boolean
  preferDistributionStepMetrics: boolean
  sendPullRequestLabels: boolean
}

export const run = async (context: GitHubContext, inputs: Inputs): Promise<void> => {
  const metricsClient = createMetricsClient(inputs)

  await handleEvent(metricsClient, context, inputs)

  const rateLimit = await getRateLimitMetrics(context, inputs)
  await metricsClient.submitMetrics(rateLimit, 'rate limit')
}

const handleEvent = async (metricsClient: MetricsClient, context: GitHubContext, inputs: Inputs) => {
  if (context.eventName === 'workflow_run') {
    return await handleWorkflowRun(metricsClient, context.payload as WorkflowRunEvent, inputs)
  }
  if (context.eventName === 'pull_request') {
    return await handlePullRequest(metricsClient, context.payload as PullRequestEvent, context, inputs)
  }
  if (context.eventName === 'push') {
    return handlePush(metricsClient, context.payload as PushEvent)
  }
  if (context.eventName === 'schedule') {
    return handleSchedule(metricsClient, context, inputs)
  }
  core.warning(`Not supported event ${context.eventName}`)
}

const getRateLimitMetrics = async (context: GitHubContext, inputs: Inputs) => {
  const octokit = github.getOctokit(inputs.githubTokenForRateLimitMetrics)
  const rateLimit = await octokit.rest.rateLimit.get()
  return computeRateLimitMetrics(context, rateLimit)
}
