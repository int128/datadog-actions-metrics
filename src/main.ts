import * as core from '@actions/core'
import * as github from '@actions/github'
import { run } from './run'

const main = async (): Promise<void> => {
  await run(github.context, {
    githubToken: core.getInput('github-token', { required: true }),
    githubTokenForRateLimitMetrics: core.getInput('github-token-rate-limit-metrics', { required: true }),
    datadogApiKey: core.getInput('datadog-api-key') || undefined,
    datadogSite: core.getInput('datadog-site') || undefined,
    datadogTags: core.getMultilineInput('datadog-tags'),
    collectJobMetrics: core.getBooleanInput('collect-job-metrics'),
    collectStepMetrics: core.getBooleanInput('collect-step-metrics'),
    preferDistributionWorkflowRunMetrics: core.getBooleanInput('prefer-distribution-workflow-run-metrics'),
    preferDistributionJobMetrics: core.getBooleanInput('prefer-distribution-job-metrics'),
    preferDistributionStepMetrics: core.getBooleanInput('prefer-distribution-step-metrics'),
    sendPullRequestLabels: core.getBooleanInput('send-pull-request-labels'),
  })
}

main().catch((e: Error) => {
  core.setFailed(e)
  console.error(e)
})
