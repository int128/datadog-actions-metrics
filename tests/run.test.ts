import * as github from '@actions/github'
import { v1 } from '@datadog/datadog-api-client'
import { run } from '../src/run'
import { exampleWorkflowRunCompletedEvent } from './fixtures'
import { exampleRateLimitResponse } from './rateLimit/fixtures'
import { exampleCompletedCheckSuite } from './workflowRun/fixtures/completedCheckSuite'
import { examplePullRequestClosedEvent } from './fixtures'
import { WebhookPayload } from '@actions/github/lib/interfaces'
import { examplePullRequestOpenedEvent } from './fixtures'
import { exampleGetPullRequestQuery } from './pullRequest/fixtures/getPullRequest'
import { exampleWorkflowJobs } from './workflowRun/fixtures/workflowJobs'

jest.mock('@actions/core')

jest.mock('@actions/github')
const octokitMock = {
  graphql: jest.fn(),
  rest: {
    actions: {
      listJobsForWorkflowRunAttempt: jest.fn(),
    },
    rateLimit: {
      get: jest.fn(),
    },
  },
}
const getOctokit = github.getOctokit as jest.Mock
getOctokit.mockReturnValue(octokitMock)

jest.mock('@datadog/datadog-api-client')
const submitMetrics = jest.spyOn(v1.MetricsApi.prototype, 'submitMetrics')

test('workflow_run with collectJobMetrics', async () => {
  octokitMock.rest.actions.listJobsForWorkflowRunAttempt.mockResolvedValue({ data: exampleWorkflowJobs })
  octokitMock.graphql.mockResolvedValue(exampleCompletedCheckSuite)
  octokitMock.rest.rateLimit.get.mockResolvedValue(exampleRateLimitResponse)
  submitMetrics.mockResolvedValue({ status: 'ok' })

  await run(
    {
      eventName: 'workflow_run',
      payload: exampleWorkflowRunCompletedEvent,
      repo: { owner: 'Codertocat', repo: 'Hello-World' },
    },
    {
      githubToken: 'GITHUB_TOKEN',
      githubTokenForRateLimitMetrics: 'GITHUB_TOKEN',
      datadogApiKey: 'DATADOG_API_KEY',
      datadogTags: [],
      collectJobMetrics: true,
      collectStepMetrics: true,
      preferDistributionWorkflowRunMetrics: false,
      preferDistributionJobMetrics: false,
      preferDistributionStepMetrics: false,
      sendPullRequestLabels: false,
    },
  )
  expect(getOctokit).toHaveBeenCalledWith('GITHUB_TOKEN')
  expect(submitMetrics).toHaveBeenCalledTimes(4)
  expect(submitMetrics.mock.calls).toMatchSnapshot()
})

test('workflow_run', async () => {
  octokitMock.rest.rateLimit.get.mockResolvedValue(exampleRateLimitResponse)
  submitMetrics.mockResolvedValue({ status: 'ok' })

  await run(
    {
      eventName: 'workflow_run',
      payload: exampleWorkflowRunCompletedEvent,
      repo: { owner: 'Codertocat', repo: 'Hello-World' },
    },
    {
      githubToken: 'GITHUB_TOKEN',
      githubTokenForRateLimitMetrics: 'GITHUB_TOKEN',
      datadogApiKey: 'DATADOG_API_KEY',
      datadogTags: [],
      collectJobMetrics: false,
      collectStepMetrics: false,
      preferDistributionWorkflowRunMetrics: false,
      preferDistributionJobMetrics: false,
      preferDistributionStepMetrics: false,
      sendPullRequestLabels: false,
    },
  )
  expect(getOctokit).toHaveBeenCalledWith('GITHUB_TOKEN')
  expect(submitMetrics).toHaveBeenCalledTimes(2)
  expect(submitMetrics.mock.calls).toMatchSnapshot()
})

test('pull_request_opened', async () => {
  octokitMock.rest.rateLimit.get.mockResolvedValue(exampleRateLimitResponse)
  submitMetrics.mockResolvedValue({ status: 'ok' })

  await run(
    {
      eventName: 'pull_request',
      payload: examplePullRequestOpenedEvent as WebhookPayload,
      repo: { owner: 'Codertocat', repo: 'Hello-World' },
    },
    {
      githubToken: 'GITHUB_TOKEN',
      githubTokenForRateLimitMetrics: 'GITHUB_TOKEN',
      datadogApiKey: 'DATADOG_API_KEY',
      datadogTags: [],
      collectJobMetrics: false,
      collectStepMetrics: false,
      preferDistributionWorkflowRunMetrics: false,
      preferDistributionJobMetrics: false,
      preferDistributionStepMetrics: false,
      sendPullRequestLabels: false,
    },
  )
  expect(getOctokit).toHaveBeenCalledWith('GITHUB_TOKEN')
  expect(submitMetrics).toHaveBeenCalledTimes(2)
  expect(submitMetrics.mock.calls).toMatchSnapshot()
})

test('pull_request_closed', async () => {
  octokitMock.graphql.mockResolvedValue(exampleGetPullRequestQuery)
  octokitMock.rest.rateLimit.get.mockResolvedValue(exampleRateLimitResponse)
  submitMetrics.mockResolvedValue({ status: 'ok' })

  await run(
    {
      eventName: 'pull_request',
      payload: examplePullRequestClosedEvent as WebhookPayload,
      repo: { owner: 'Codertocat', repo: 'Hello-World' },
    },
    {
      githubToken: 'GITHUB_TOKEN',
      githubTokenForRateLimitMetrics: 'GITHUB_TOKEN',
      datadogApiKey: 'DATADOG_API_KEY',
      datadogTags: [],
      collectJobMetrics: false,
      collectStepMetrics: false,
      preferDistributionWorkflowRunMetrics: false,
      preferDistributionJobMetrics: false,
      preferDistributionStepMetrics: false,
      sendPullRequestLabels: true,
    },
  )
  expect(getOctokit).toHaveBeenCalledWith('GITHUB_TOKEN')
  expect(submitMetrics).toHaveBeenCalledTimes(2)
  expect(submitMetrics.mock.calls).toMatchSnapshot()
})
