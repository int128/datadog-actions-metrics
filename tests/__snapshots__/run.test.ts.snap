// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pull_request_closed 1`] = `
Array [
  Object {
    "body": Object {
      "series": Array [
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.total",
          "points": Array [
            Array [
              1558279233,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.since_opened_seconds",
          "points": Array [
            Array [
              1558279233,
              345600,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.commits",
          "points": Array [
            Array [
              1558279233,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.changed_files",
          "points": Array [
            Array [
              1558279233,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.additions",
          "points": Array [
            Array [
              1558279233,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.deletions",
          "points": Array [
            Array [
              1558279233,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.since_first_authored_seconds",
          "points": Array [
            Array [
              1558279233,
              346822,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.since_first_committed_seconds",
          "points": Array [
            Array [
              1558279233,
              346151,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_closed.first_review_time_seconds",
          "points": Array [
            Array [
              1558279233,
              600,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
            "merged:false",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.remaining",
          "points": Array [
            Array [
              1445412480,
              3999,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:core",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.limit",
          "points": Array [
            Array [
              1445412480,
              5000,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:core",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.remaining",
          "points": Array [
            Array [
              1445412480,
              999,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:search",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.limit",
          "points": Array [
            Array [
              1445412480,
              6000,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:search",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.remaining",
          "points": Array [
            Array [
              1445412480,
              4999,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:graphql",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.limit",
          "points": Array [
            Array [
              1445412480,
              7000,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:graphql",
          ],
          "type": "gauge",
        },
      ],
    },
  },
]
`;

exports[`pull_request_opened 1`] = `
Array [
  Object {
    "body": Object {
      "series": Array [
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_opened.total",
          "points": Array [
            Array [
              1557933633,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_opened.commits",
          "points": Array [
            Array [
              1557933633,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_opened.changed_files",
          "points": Array [
            Array [
              1557933633,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_opened.additions",
          "points": Array [
            Array [
              1557933633,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.pull_request_opened.deletions",
          "points": Array [
            Array [
              1557933633,
              1,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "sender:Codertocat",
            "sender_type:User",
            "user:Codertocat",
            "pull_request_number:2",
            "draft:false",
            "base_ref:master",
            "head_ref:changes",
          ],
          "type": "count",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.remaining",
          "points": Array [
            Array [
              1445412480,
              3999,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:core",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.limit",
          "points": Array [
            Array [
              1445412480,
              5000,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:core",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.remaining",
          "points": Array [
            Array [
              1445412480,
              999,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:search",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.limit",
          "points": Array [
            Array [
              1445412480,
              6000,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:search",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.remaining",
          "points": Array [
            Array [
              1445412480,
              4999,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:graphql",
          ],
          "type": "gauge",
        },
        Object {
          "host": "github.com",
          "metric": "github.actions.api_rate_limit.limit",
          "points": Array [
            Array [
              1445412480,
              7000,
            ],
          ],
          "tags": Array [
            "repository_owner:Codertocat",
            "repository_name:Hello-World",
            "resource:graphql",
          ],
          "type": "gauge",
        },
      ],
    },
  },
]
`;
